### ISSUE TITLE
[gym] Export Developer ID macOS app fails (.app)

### ISSUE THREAD

**Issue by EpaL** [fastlane/fastlane#15963]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
Note: I'm a little new to fastlane so please bare with me!

Using `gym` to build my macOS app, the command fails right after the export phase:

```
INFO [2020-02-02 18:34:26.67]: ▸ Exported MyApp to: /var/folders/sr/4s52t1q16yzgnn17nfwy_fv40000gn/T/gym_output20200202-85761-10rcbbx
INFO [2020-02-02 18:34:26.67]: ▸ ** EXPORT SUCCEEDED **
ERROR [2020-02-02 18:34:26.69]: The generated archive is invalid, this can have various reasons:
ERROR [2020-02-02 18:34:26.69]: Usually it's caused by the `Skip Install` option in Xcode, set it to `NO`
ERROR [2020-02-02 18:34:26.69]: For more information visit https://developer.apple.com/library/ios/technotes/tn2215/_index.html
ERROR [2020-02-02 18:34:26.69]: Also, make sure to have a valid code signing identity and provisioning profile installed
...
```

The export itself succeeds (a valid, correctly signed binary is produced) but fastlane appears unable to find the binary and assumes the export has failed.

This appears to be a bug in `gym/lib/gym/generators/package_command_generator_xcode7.rb` in `pkg_path`. The function assumes the export output has a suffix of `.pkg`. However, I am trying to export a Developer ID app, so the export produces a `.app`. 

If I 'hack' the `pkg_path` function and replace `.pkg` with `.app`, everything works perfectly.

##### Command executed
bundle exec fastlane myappbeta

##### Complete output when running fastlane, including the stack trace and command used
<details>
  <pre> [18:41:29]: Successfully exported and compressed dSYM file
[18:41:29]: Successfully exported the .app file:
[18:41:29]: ../MyApp Releases/MyApp/MyApp 4.1.10/MyApp 4.app
[18:41:29]: Generated plist file with the following values:
[18:41:29]: ▸ -----------------------------------------
[18:41:29]: ▸ {
[18:41:29]: ▸   "provisioningProfiles": {
[18:41:29]: ▸     "com.xxx.myapp": "match Direct com.xxx.myapp macos"
[18:41:29]: ▸   },
[18:41:29]: ▸   "method": "developer-id",
[18:41:29]: ▸   "signingStyle": "manual"
[18:41:29]: ▸ }
[18:41:29]: ▸ -----------------------------------------
[18:41:29]: $ /usr/bin/xcrun /Users/blah/Work/Xcode\ Projects/MyApp/vendor/bundle/ruby/2.6.0/gems/fastlane-2.141.0/gym/lib/assets/wrap_xcodebuild/xcbuild-safe.sh -exportArchive -exportOptionsPlist '/var/folders/sr/4s52t1q16yzgnn17nfwy_fv40000gn/T/gym_config20200202-95104-14e0ysk.plist' -archivePath /Users/blah/Library/Developer/Xcode/Archives/2020-02-02/MyApp\ 4\ 2020-02-02\ 18.41.09.xcarchive -exportPath '/var/folders/sr/4s52t1q16yzgnn17nfwy_fv40000gn/T/gym_output20200202-95104-1vwto0x'
[18:41:36]: The generated archive is invalid, this can have various reasons:
[18:41:36]: Usually it's caused by the `Skip Install` option in Xcode, set it to `NO`
[18:41:36]: For more information visit https://developer.apple.com/library/ios/technotes/tn2215/_index.html
[18:41:36]: Also, make sure to have a valid code signing identity and provisioning profile installed
[18:41:36]: Follow this guide to setup code signing https://docs.fastlane.tools/codesigning/GettingStarted/
[18:41:36]: If your intention was only to export an ipa be sure to provide a valid archive at the archive path.
[18:41:36]: This error might also happen if your workspace/project file is not in the root directory of your project.
[18:41:36]: To workaround that issue, you can wrap your calls to gym with
[18:41:36]: `Dir.chdir('../path/to/dir/containing/proj') do`
[18:41:36]: For an example you can check out
[18:41:36]: https://github.com/artsy/emission-nebula/commit/44fe51a7fea8f7d52f0f77d6c3084827fe5dd59e </pre>
</details>

### Environment

<details>
  <pre> <details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                                |
| --------------------------- | -------------------------------------------------------------------- |
| OS                          | 10.15.3                                                              |
| Ruby                        | 2.6.3                                                                |
| Bundler?                    | true                                                                 |
| Git                         | git version 2.21.1 (Apple Git-122.3)                                 |
| Installation Source         | ~/Work/Xcode Projects/MyApp/vendor/bundle/ruby/2.6.0/bin/fastlane |
| Host                        | Mac OS X 10.15.3 (19D76)                                             |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib       |
| OpenSSL Version             | LibreSSL 2.8.3                                                       |
| Is contained                | false                                                                |
| Is homebrew                 | false                                                                |
| Is installed via Fabric.app | false                                                                |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                          |
| Xcode Version               | 11.3.1                                                               |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# This file contains the fastlane.tools configuration
# You can find the documentation at https://docs.fastlane.tools
#
# For a list of all available actions, check out
#
#     https://docs.fastlane.tools/actions
#
# For a list of all available plugins, check out
#
#     https://docs.fastlane.tools/plugins/available-plugins
#

# Uncomment the line if you want fastlane to automatically update itself
# update_fastlane

# load variables from .env file in the root if it exists
if File.exist?('../.env')
  open('../.env', 'r').readlines.each do |l|
    kv = l.split('=')
    ENV[kv[0]] = kv[1].chomp
  end
end

default_platform(:mac)

platform :mac do
  desc "MyApp"

  # Configuration
  projectFilename = "MyApp.xcodeproj"

  # Before Integration Tests (Xcode Server Bot)
  lane :before_integration do
    sigh(platform: "macos",
         output_path: '/Library/Developer/XcodeServer/ProvisioningProfiles', 
         skip_install: true)
  end

  # Generate Beta and export it to Latest folder
  lane :myappbeta do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppPaddle"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/Paddle/" + versionedPath
    zipFilename = versionedFullPath + "/MyApp Paddle " + version + "-" + buildNumber + ".zip"
    appFilename = versionedFullPath + "/MyApp 4.app"

    # Set up certificates and provisioning profiles
    match(
      platform: "macos",
      type: "developer_id"
    )

    # Note: To make this work, currently need to edit /Users/blah/Work/Xcode Projects/MyApp/vendor/bundle/ruby/2.6.0/gems/fastlane-2.141.0/gym/lib/gym/generators/package_command_generator_xcode7.rb
    # around line 102 and replace all ".pkg" with ".app"
    build_mac_app(scheme: "MyApp Paddle Beta",
      output_directory: versionedFullPath,
      export_method: "developer-id",
      )

    notarize(
      package: appFilename,
      print_log: 'true',
      verbose: 'true'
    )      
      
    # Compress to a zip file
    zip(path: versionedFullPath + "/MyApp 4.app",
        output_path: zipFilename
    )
  end

  # Generate MAS Beta and export it to Latest folder
  lane :masbeta do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppMAS"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/MAS/" + versionedPath
    zipFilename = versionedFullPath + "/MyApp MAS " + version + "-" + buildNumber + ".zip"
    appFilename = versionedFullPath + "/MyApp 4.app"

    # Set up certificates and provisioning profiles
    match(
      platform: "macos",
      type: "developer_id"
    )
    build_mac_app(scheme: "MyApp MAS Beta",
        output_directory: versionedFullPath,
        export_method: "developer-id",
      )

    notarize(
      package: appFilename,
      print_log: 'true',
      verbose: 'true'
    )      
  
    # Compress to a zip file
    zip(path: versionedFullPath + "/MyApp 4.app",
        output_path: zipFilename
    )
  end  

  # Build the Paddle release version, export it and create the Disk Image distribution
  lane :buildpaddle do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppPaddle"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/Paddle/" + versionedPath
    appFilename = versionedFullPath + "/MyApp 4.app"
    zipFilename = versionedFullPath + "/MyApp " + version + ".zip"
    dmgFilename = versionedFullPath + "/MyApp " + version + ".dmg"
    dsymFilename = versionedFullPath + "/MyApp 4.app.dSYM.zip"

    # Set up certificates and provisioning profiles
    match(
      platform: "macos",
      type: "developer_id"
    )

    # Build
    gym(scheme: "MyApp Paddle",
        output_directory: versionedFullPath,
        export_method: "developer-id",
        # export_options: {
        #   method: "developer-id",
        #   provisioningProfiles: { 
        #     "com.xxx.myapp" => "Direct_com.xxx.myapp.provisionprofile"
        #   },
        # },
        # export_xcargs: "-allowProvisioningUpdates"
      )
    
    # Unfortunately, Fastlane doesn't properly export the built product (it simply copies it from the xcarchive) and it doesn't
    # use the provided provisioning profile for signing.
    # UI.important "\n\nUnfortunately, right now, this is as far as Fastlane can go.\nNext step is to visit the Organiser, Notarise the app and export the notarised copy to " +appFilename + ".\nOnce that is done, run the paddlerelease lane to continue with distribution."
    # Notarize the app
    notarize(
      package: appFilename,
      print_log: 'true',
      verbose: 'true'
    )      
  end

  # Peform the release of the Paddle version. Notarize, ZIP, upload to AppCenter, DropDMG and upload to xxx.
  lane :releasepaddle do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppPaddle"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/Paddle/" + versionedPath
    appFilename = versionedFullPath + "/MyApp 4.app"
    zipFilename = versionedFullPath + "/MyApp " + version + ".zip"
    dmgFilename = versionedFullPath + "/MyApp " + version + ".dmg"
    dsymFilename = versionedFullPath + "/MyApp 4.app.dSYM.zip"
    
    # Compress to a zip file
    begin
      # Put this item in an error handler because it may fail (and otherwise cause execution to stop)
      sh("rm", "../" + zipFilename)
    rescue => ex
      UI.error(ex)
    end    
    zip(path: appFilename,
        output_path: zipFilename
    )

    # Upload to MS AppCenter
    appcenter_upload(
      api_token: ENV["APPCENTER_API_TOKEN"],
      owner_name: ENV["APPCENTER_OWNER_NAME"],
      app_name: ENV["APPCENTER_MACOS_APP_NAME_PADDLE"],
      app_os: 'MacOS',  
      file: appFilename,
      dsym: dsymFilename,
    )    

    # Generate custom Disk Image
    begin
      # Put this item in an error handler because it may fail (and otherwise cause execution to stop)
      sh("rm", "../" + dmgFilename)
    rescue => ex
      UI.error(ex)
    end    
    sh("dropdmg", 
       "--license-folder", 
       "./DropDMG/Default", 
       "--layout-folder", 
       "DropDMG", 
       "--config-name", 
       "MyApp 4", 
       "--base-name", 
       versionedPath, 
       "../" + versionedFullPath + "/MyApp 4.app"
    )

    # Upload disk image to xxx.xxx.com
    scp(host: "xxx.xxx.com",
        username: "username",
        upload: {
          src: dmgFilename,
          dst: "/var/www/xxx/releases"
        }
    )
    # Upload ZIP to xxx.xxx.com
    scp(host: "xxx.xxx.com",
        username: "username",
        upload: {
          src: zipFilename,
          dst: "/var/www/xxx/releases"
        }
    )
  end

  # Generate MAS Beta and export it to Latest folder
  lane :masrelease do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppMAS"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/MAS/" + versionedPath
    zipFilename = versionedFullPath + "/MyApp MAS " + version + "-" + buildNumber + ".zip"
    dsymFilename = versionedFullPath + "/MyApp 4.app.dSYM.zip"
    appFilename = versionedFullPath + "/MyApp 4.app"

    # Set up certificates and provisioning profiles
    match(
      platform: "macos",
      type: "appstore"
    )

    build_mac_app(scheme: "MyApp MAS",
      output_directory: versionedFullPath,
      # export_options: {
      #   method: "package",
      #   provisioningProfiles: { 
      #     "com.xxx.myapp" => "AppStore_com.xxx.myapp.provisionprofile"
      #   }
      # },
      # export_xcargs: "-allowProvisioningUpdates"
    )

    ### Note, Gym does not yet support exporting a .pkg required for uploading to the AppStore.
    ### Once Gym is complete, go to the Xcode Organizer and submit the build to the App Store.
    ### Keep an eye on this PR that adds support for Mac Pkg and AppStore uploads:
    ### https://github.com/fastlane/fastlane/pull/12195
    deliver(
      platform: "osx",
      skip_screenshots: "true",
      skip_metadata: "true",
      run_precheck_before_submit: "false",
      force: "true"
    )      

    # Upload to MS AppCenter
    appcenter_upload(
      api_token: ENV["APPCENTER_API_TOKEN"],
      owner_name: ENV["APPCENTER_OWNER_NAME"],
      app_name: ENV["APPCENTER_MACOS_APP_NAME_MAS"],
      app_os: 'MacOS',  
      file: appFilename,
      dsym: dsymFilename,
    )    
  end  

end

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
app_identifier("com.xxx.myapp") # The bundle identifier of your app
apple_id("emailaddress") # Your Apple email address

# For more information about the Appfile, see:
#     https://docs.fastlane.tools/advanced/#appfile

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.141.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                    | Version | Update-Status |
| ------------------------- | ------- | ------------- |
| fastlane-plugin-appcenter | 1.7.1   | ✅ Up-To-Date  |
| fastlane-plugin-notarize  | 1.1.2   | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.3.0        |
| bundler                   | 1.17.2       |
| etc                       | 1.0.1        |
| forwardable               | 1.2.0        |
| CFPropertyList            | 3.0.2        |
| public_suffix             | 2.0.5        |
| addressable               | 2.7.0        |
| atomos                    | 0.1.3        |
| babosa                    | 1.0.3        |
| claide                    | 1.0.3        |
| colored                   | 1.2          |
| colored2                  | 3.1.2        |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| digest-crc                | 0.4.1        |
| unf_ext                   | 0.0.7.6      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| dotenv                    | 2.7.5        |
| emoji_regex               | 1.0.1        |
| excon                     | 0.72.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 0.17.3       |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.13.1       |
| fastimage                 | 2.1.7        |
| gh_inspector              | 1.1.3        |
| jwt                       | 2.1.0        |
| memoist                   | 0.16.2       |
| multi_json                | 1.14.1       |
| os                        | 1.0.1        |
| signet                    | 0.12.0       |
| googleauth                | 0.10.0       |
| httpclient                | 2.8.3        |
| mini_mime                 | 1.0.2        |
| uber                      | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| google-api-client         | 0.36.4       |
| google-cloud-env          | 1.3.0        |
| google-cloud-errors       | 1.0.0        |
| google-cloud-core         | 1.5.0        |
| google-cloud-storage      | 1.25.1       |
| json                      | 2.3.0        |
| mini_magick               | 4.10.1       |
| multi_xml                 | 0.6.0        |
| plist                     | 3.5.0        |
| rubyzip                   | 1.3.0        |
| security                  | 0.1.3        |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.7        |
| slack-notifier            | 2.3.2        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.6.1        |
| terminal-table            | 1.8.0        |
| tty-screen                | 0.7.0        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| word_wrap                 | 1.0.0        |
| nanaimo                   | 0.2.6        |
| xcodeproj                 | 1.14.0       |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| xcpretty-travis-formatter | 1.0.0        |
| fastlane-plugin-appcenter | 1.7.1        |
| fastlane-plugin-notarize  | 1.1.2        |
| net-ssh                   | 5.2.0        |
| net-scp                   | 2.0.0        |
</details>

*generated on:* **2020-02-02**
</details> </pre>
</details>

**Comment by ljc-appest** [comment]
same issue here

**Comment by ljc-appest** [comment]
```
ERROR [2020-02-07 20:02:48.34]: The generated archive is invalid, this can have various reasons:
ERROR [2020-02-07 20:02:48.34]: Usually it's caused by the `Skip Install` option in Xcode, set it to `NO`
ERROR [2020-02-07 20:02:48.34]: For more information visit https://developer.apple.com/library/ios/technotes/tn2215/_index.html
ERROR [2020-02-07 20:02:48.34]: Also, make sure to have a valid code signing identity and provisioning profile installed
ERROR [2020-02-07 20:02:48.34]: Follow this guide to setup code signing https://docs.fastlane.tools/codesigning/GettingStarted/
ERROR [2020-02-07 20:02:48.34]: If your intention was only to export an ipa be sure to provide a valid archive at the archive path.
ERROR [2020-02-07 20:02:48.34]: This error might also happen if your workspace/project file is not in the root directory of your project.
ERROR [2020-02-07 20:02:48.34]: To workaround that issue, you can wrap your calls to gym with
ERROR [2020-02-07 20:02:48.36]: `Dir.chdir('../path/to/dir/containing/proj') do`
ERROR [2020-02-07 20:02:48.36]: For an example you can check out
ERROR [2020-02-07 20:02:48.36]: https://github.com/artsy/emission-nebula/commit/44fe51a7fea8f7d52f0f77d6c3084827fe5dd59e
WARN [2020-02-07 20:02:48.36]: Lane Context:
INFO [2020-02-07 20:02:48.36]: {:DEFAULT_PLATFORM=>:mac, :PLATFORM_NAME=>:mac, :LANE_NAME=>"mac outside", :SIGH_PROFILE_PATH=>"/Users/adad184/Workspace/mac/Build/Mac Outside widget.mobileprovision", :SIGH_PROFILE_PATHS=>["/Users/adad184/Workspace/mac/Build/Mac Outside app.mobileprovision", "/Users/xx/Workspace/mac/Build/Mac Outside widget.mobileprovision"], :SIGH_UDID=>"bf3ab3db-ead9-4486-91d8-b04bdb740c19", :SIGH_UUID=>"bf3ab3db-ead9-4486-91d8-b04bdb740c19", :SIGH_NAME=>"Mac Outside com.xx.task.mac.xx-Today-Widget", :SIGH_PROFILE_TYPE=>"developer-id"}
ERROR [2020-02-07 20:02:48.36]: Archive invalid
```

I got this when I build outside pkg, and it's always show "**Archive invalid**" without any detail.

I can build appstore pkg successfully, with the same fastfile with different lane

**Comment by janpio** [comment]
@ljc-appest Please open a new issue for that - and fill the complete issue template. Please also include a description what you mean by "outside pkg" in there. Thanks.

@EpaL Interesting, indeed looks like a bug (or missing feature to support `.app` archives). This is quite new code from @joshdholtz, so he probably just missed this variant. Thanks for reporting!

Should we just look for `.pkg` _and_ `.app` here or is there a better way to know what we should look for?

**Comment by EpaL** [comment]
Thanks @janpio. That's a good question, I'm not familiar enough with the various export methods to know if there are any other file extensions. For macOS exports, I suspect .pkg, .app and perhaps .mpkg will cover them all but don't take my word for it.

**Comment by waleeg** [comment]
👋 Can confirm,  same error. Unable to build anymore for macOS (AppKit .app). Probably since the Fastlane update to support Catalyst.

**Comment by khallmark** [comment]
Confirming the same error. Can't export Developer ID signed application.

**Comment by weevilgenius** [comment]
I had the same issue: Exporting a Developer ID signed Mac OS application fails with a mysterious "archive invalid" message.  It turns out that Fastlane 2.140.0 introduced a change to gym (see #16073) which causes it to always try to export a .pkg file instead of .app. Adding `skip_package_pkg: true` restores the old behavior and fixes the issue.

**Comment by EpaL** [comment]
> I had the same issue: Exporting a Developer ID signed Mac OS application fails with a mysterious "archive invalid" message. It turns out that Fastlane 2.140.0 introduced a change to gym (see #16073) which causes it to always try to export a .pkg file instead of .app. Adding `skip_package_pkg: true` restores the old behavior and fixes the issue.

Actually, this workaround also has a rather nasty unintended side-effect of breaking exporting. Adding `skip_package_pkg: true` fixes the original error, but it then no longer processes any of the export options, rendering it basically useless for packaging macOS apps for distribution.

Seems this may need a proper fix after all...

**Comment by someyura** [comment]
Confirming the same error. Still having the same problem on Fastlane 2.148.1

**Comment by trondkr** [comment]
Confirming the same error. Still having the same problem on Fastlane 2.150.3

**Comment by eliottha** [comment]
Any update on this?

**Comment by khallmark** [comment]
> > I had the same issue: Exporting a Developer ID signed Mac OS application fails with a mysterious "archive invalid" message. It turns out that Fastlane 2.140.0 introduced a change to gym (see #16073) which causes it to always try to export a .pkg file instead of .app. Adding `skip_package_pkg: true` restores the old behavior and fixes the issue.
> 
> Actually, this workaround also has a rather nasty unintended side-effect of breaking exporting. Adding `skip_package_pkg: true` fixes the original error, but it then no longer processes any of the export options, rendering it basically useless for packaging macOS apps for distribution.
> 
> Seems this may need a proper fix after all...

Archive your app. Export it with developer Id signing. Take the exportOptions.plist and use that with fastlane.

**Comment by joshdholtz** [comment]
@trondkr @eliottha Does setting `export_method: ‘developer-id’` and `skip_package_pkg: true` work for you? If so, I can add a fix this week to do that by default for developer ID 💪

**Comment by weevilgenius** [comment]
I can't speak for others, but we use `skip_package_pkg: true` and `export_options: fastlane/ExportOptions.plist` and it works correctly. Our app passes notarization, so all signing appears to be correct.

**Comment by trondkr** [comment]
@weevilgenius Would it be possible to see what your `fastlane/ExportOptions.plist` looks like? I am still struggling to get this tp work correctly.

**Comment by waleeg** [comment]
> @trondkr @eliottha Does setting `export_method: ‘developer-id’` and `skip_package_pkg: true` work for you? If so, I can add a fix this week to do that by default for developer ID 💪

It worked for me.

**Comment by weevilgenius** [comment]
@trondkr I used the one generated from Xcode when exporting the app. The options you select during the export change what goes into the file.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>destination</key>
	<string>export</string>
	<key>method</key>
	<string>developer-id</string>
	<key>signingCertificate</key>
	<string>286A48D7AB9A32F47FE5DF8775A154</string>
	<key>signingStyle</key>
	<string>manual</string>
	<key>teamID</key>
	<string>xxxxxxx</string>
</dict>
</plist>
```

**Comment by EpaL** [comment]
> I can't speak for others, but we use `skip_package_pkg: true` and `export_options: fastlane/ExportOptions.plist` and it works correctly. Our app passes notarization, so all signing appears to be correct.

Unfortunately it doesn't work for me.

Using:
```
    build_mac_app(scheme: "MyApp",
    export_options: "fastlane/ExportOptions.plist")
```

... and the following ExportOptions.plist:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>destination</key>
	<string>export</string>
	<key>method</key>
	<string>developer-id</string>
	<key>signingStyle</key>
	<string>automatic</string>
	<key>teamID</key>
	<string>xxxxxxxxxx</string>
</dict>
</plist>
```

It builds the app and starts the export, but the export and signing options are wrong (not the same as the plist above):

```
[19:38:01]: Successfully exported and compressed dSYM file
[19:38:01]: Successfully exported the .app file:
[19:38:01]: ../MyApp/Latest/MyApp.app
[19:38:01]: Generated plist file with the following values:
[19:38:01]: ▸ -----------------------------------------
[19:38:01]: ▸ {
[19:38:01]: ▸   "destination": "export",
[19:38:01]: ▸   "method": "app-store",
[19:38:01]: ▸   "signingStyle": "automatic",
[19:38:01]: ▸   "teamID": "xxxxxxxx",
[19:38:01]: ▸   "installerSigningCertificate": "3rd Party Mac Developer Installer: Me (xxxxxxxxxx)"
[19:38:01]: ▸ }
[19:38:01]: ▸ -----------------------------------------
```

If I use the same ExportOptions.plist with the following arguments:

```
    build_mac_app(scheme: "MyApp",
      export_options: "fastlane/ExportOptions.plist",
      skip_package_pkg: true
    )

```
```
[19:42:23]: ▸ Archive Succeeded
[19:42:24]: Compressing 2 dSYM(s)
[19:42:24]: $ cd '/Users/xx/Library/Developer/Xcode/Archives/2020-07-28/MyApp 2020-07-28 19.42.00.xcarchive/dSYMs' && zip -r '/Users/xx/Work/Xcode Projects/MyApp Releases/Latest/MyApp.app.dSYM.zip' *.dSYM
[19:42:24]: ▸ updating: MyApp.app.dSYM/ (stored 0%)
[19:42:24]: ▸ updating: MyApp.app.dSYM/Contents/ (stored 0%)
[19:42:24]: ▸ updating: MyApp.app.dSYM/Contents/Resources/ (stored 0%)
[19:42:24]: ▸ updating: MyApp.app.dSYM/Contents/Resources/DWARF/ (stored 0%)
[19:42:24]: ▸ updating: MyApp.app.dSYM/Contents/Resources/DWARF/MyApp (deflated 69%)
[19:42:24]: ▸ updating: MyApp.app.dSYM/Contents/Info.plist (deflated 51%)
[19:42:24]: ▸ updating: Sparkle.framework.dSYM/ (stored 0%)
[19:42:24]: ▸ updating: Sparkle.framework.dSYM/Contents/ (stored 0%)
[19:42:24]: ▸ updating: Sparkle.framework.dSYM/Contents/Resources/ (stored 0%)
[19:42:24]: ▸ updating: Sparkle.framework.dSYM/Contents/Resources/DWARF/ (stored 0%)
[19:42:24]: ▸ updating: Sparkle.framework.dSYM/Contents/Resources/DWARF/Sparkle (deflated 63%)
[19:42:24]: ▸ updating: Sparkle.framework.dSYM/Contents/Info.plist (deflated 52%)

[19:42:24]: Successfully exported and compressed dSYM file
[19:42:25]: Successfully exported the .app file:
[19:42:25]: ../MyApp Releases/Latest/MyApp.app
+------------------------------------+-----------------------------------------------------------------------------+
|                                                   Lane Context                                                   |
+------------------------------------+-----------------------------------------------------------------------------+
| DEFAULT_PLATFORM                   | mac                                                                         |
| PLATFORM_NAME                      | Mac                                                                         |

...
```

... it skips signing all together and spits out a developer-signed binary.

Not sure what I'm doing wrong here... any help much appreciated!

**Comment by joshdholtz** [comment]
@EpaL Can you try this? 👇 

```rb
build_mac_app(
  scheme: "MyApp",
  export_method: 'developer-id',
  skip_package_pkg: true
)
```

**Comment by EpaL** [comment]
@joshdholtz thanks. It produced an identical result to `build_mac_app(scheme: "MyApp",
      export_options: "fastlane/ExportOptions.plist",
      skip_package_pkg: true
    )` above. That is, it builds but then it does not attempt to sign the .app.

**Comment by joshdholtz** [comment]
@EpaL Your _fastlane_ output from above looks similar to mine before I pass it to the `notarize ` action 🤔 Are you getting an error from `build_mac_app` or the `notarize` action about signing not being right?

**Comment by EpaL** [comment]
It's the `notarize` action that's returning the error:

```
[21:46:02]: Successfully exported and compressed dSYM file
[21:46:03]: Successfully exported the .app file:
[21:46:03]: ../MyApp Releases/Latest/MyApp.app
[21:46:03]: ----------------------
[21:46:03]: --- Step: notarize ---
[21:46:03]: ----------------------
[21:46:03]: $ ditto -c -k --rsrc --keepParent "../MyApp Releases/Latest/MyApp.app" "../MyApp Releases/Latest/MyApp.app.zip"
[21:46:04]: $ /usr/libexec/PlistBuddy -c "Print :CFBundleIdentifier" "../MyApp Releases/Latest/MyApp.app/Contents/Info.plist"
[21:46:04]: ▸ com.myapp.myapp
[21:46:04]: Uploading package to notarization service, might take a while
[21:46:04]: $ xcrun altool --notarize-app -t osx -f "../MyApp Releases/Latest/MyApp.app.zip" --primary-bundle-id com.myapp.myapp -u user@domain.com -p @env:FL_NOTARIZE_PASSWORD --output-format xml
... a bunch of XML ...
[21:47:55]: Querying request status
[21:47:55]: $ xcrun altool --notarization-info xx -u user@domain.com -p @env:FL_NOTARIZE_PASSWORD --output-format xml
[21:47:56]: ▸ <?xml version="1.0" encoding="UTF-8"?>
[21:47:56]: ▸ <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
[21:47:56]: ▸ <plist version="1.0">
[21:47:56]: ▸ <dict>
[21:47:56]: ▸ <key>notarization-info</key>
[21:47:56]: ▸ <dict>
[21:47:56]: ▸ <key>Date</key>
[21:47:56]: ▸ <date>2020-07-29T01:46:22Z</date>
[21:47:56]: ▸ <key>Hash</key>
[21:47:56]: ▸ <string>xx</string>
[21:47:56]: ▸ <key>LogFileURL</key>
[21:47:56]: ▸ <string>https://osxapps-ssl.itunes.apple.com/itunes-assets/Enigma114//developer_log.json?accessKey=xx</string>
[21:47:56]: ▸ <key>RequestUUID</key>
[21:47:56]: ▸ <string>xx</string>
[21:47:56]: ▸ <key>Status</key>
[21:47:56]: ▸ <string>invalid</string>
[21:47:56]: ▸ <key>Status Code</key>
[21:47:56]: ▸ <integer>2</integer>
[21:47:56]: ▸ <key>Status Message</key>
[21:47:56]: ▸ <string>Package Invalid</string>
[21:47:56]: ▸ </dict>
[21:47:56]: ▸ <key>os-version</key>
[21:47:56]: ▸ <string>10.15.6</string>
[21:47:56]: ▸ <key>success-message</key>
[21:47:56]: ▸ <string>No errors getting notarization info.</string>
[21:47:56]: ▸ <key>tool-path</key>
[21:47:56]: ▸ <string>/Applications/Xcode.app/Contents/SharedFrameworks/ContentDeliveryServices.framework/Versions/A/Frameworks/AppStoreService.framework</string>
[21:47:56]: ▸ <key>tool-version</key>
[21:47:56]: ▸ <string>4.01.1182</string>
[21:47:56]: ▸ </dict>
[21:47:56]: ▸ </plist>
+------------------------------------+-----------------------------------------------------------------------------+
|                                                   Lane Context                                                   |
+------------------------------------+-----------------------------------------------------------------------------+
| DEFAULT_PLATFORM                   | mac                                                                         |
| PLATFORM_NAME                      | mac                                                                         |
| LANE_NAME                          | mac myappbeta                                                              |
| GET_INFO_PLIST_VALUE_CUSTOM_VALUE  | 36789                                                                       |
| VERSION_NUMBER                     | 4.2.0-beta                                                                  |
| SIGH_PROFILE_TYPE                  | app-store                                                                   |
| MATCH_PROVISIONING_PROFILE_MAPPING | {"com.myapp.myapp"=>"match Direct com.myapp.myapp macos"} |
| XCODEBUILD_ARCHIVE                 | /Users/xx/Library/Developer/Xcode/Archives/2020-07-28/MyApp      |
|                                    | 2020-07-28 21.45.38.xcarchive                                               |
+------------------------------------+-----------------------------------------------------------------------------+
[21:47:57]: Could not notarize package with message 'Package Invalid', with log:
{
  "logFormatVersion": 1,
  "jobId": "10026b50-bb5f-4aba-a674-1c38f40c8b48",
  "status": "Invalid",
  "statusSummary": "Archive contains critical validation errors",
  "statusCode": 4000,
  "archiveFilename": "MyApp.app.zip",
  "uploadDate": "2020-07-29T01:46:22Z",
  "sha256": "27fb86e6f3c1e137d0883dc2f3dbb4f7b19904f5d035e2c4fbdcb90778355de8",
  "ticketContents": null,
  "issues": [
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.zip/MyApp.app/Contents/MacOS/MyApp",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.app.zip/MyApp.app/Contents/MacOS/MyApp",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.app.zip/MyApp .app/Contents/Library/LoginItems/LaunchAtLoginHelper.app/Contents/MacOS/LaunchAtLoginHelper",
      "message": "The binary is not signed with a valid Developer ID certificate.",
      "docUrl": null,
      "architecture": "x86_64"
    },
...
```

**Comment by khallmark** [comment]
I manually handle the signing/profiles myself due to some entitlements we have. Our ExportOptions.plist looks like this and specifies both the certificate and the profile. This file was automatically generated as part of an export.

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>destination</key>
	<string>export</string>
	<key>method</key>
	<string>developer-id</string>
	<key>provisioningProfiles</key>
	<dict>
		<key>com.mybundleid.app</key>
		<string>macOS Production Profile</string>
	</dict>
	<key>signingCertificate</key>
	<string>406869ECA3FC8....</string>
	<key>signingStyle</key>
	<string>manual</string>
	<key>teamID</key>
	<string>J6DC9...</string>
</dict>
</plist>

```

**Comment by EpaL** [comment]
I figured out the problem. I've been relying on the export process to resign my app with a Developer ID profile before notarization. Setting `skip_package_pkg: true` allows me to get around the `gym` issue where the output is a .app rather than a .pkg and gym thinks the archive process failed, but it does that by skipping the export process all together, which was how I was ensuring my app is correctly signed.

To work around this, I have gone back from Xcode Automatic Signing to Manual Signing in order to produce a Developer ID-signed binary right from the build, without requiring an export.

**Comment by webmasterdragon** [comment]
Same issue here. How can we export signed package?
Above work around is only working `skip_package_pkg: true`, but I need .pkg output.

**Comment by algentok** [comment]
> Same issue here. How can we export signed package?
> Above work around is only working `skip_package_pkg: true`, but I need .pkg output.

Probably not the best workaround, but it works for me: 
I disabled pkg creation, then executed following:
pkgbuild ----analyze --root ./MyApp.app MyApp.plist
pkgbuild --root ./MyApp.app --component-plist MyApp.plist --identifier "com.MyApp.MyApp"

**Comment by michaellenaghan** [comment]
I'm running into the same problem, though my situation might be a bit different. I'm trying to automate `development` builds of a Catalyst app. I want the app to use the `Production` iCloud environment.

This:

```ruby
    build_mac_app(
        export_options: {
            method: 'development',
            iCloudContainerEnvironment: 'Production'
        }
    )
```

produces this error:

```
[18:59:40]: ▸ Archive Succeeded
[18:59:40]: Compressing 1 dSYM(s)
[18:59:40]: $ cd '/Users/.../Library/Developer/Xcode/Archives/2020-09-30/ExampleApp 2020-09-30 18.56.39.xcarchive/dSYMs' && zip -r '/Users/.../ExampleApp/fastlane/builds/gamma/ExampleApp.app.dSYM.zip' *.dSYM
[18:59:41]: ▸ updating: ExampleApp.app.dSYM/ (stored 0%)
[18:59:41]: ▸ updating: ExampleApp.app.dSYM/Contents/ (stored 0%)
[18:59:41]: ▸ updating: ExampleApp.app.dSYM/Contents/Resources/ (stored 0%)
[18:59:41]: ▸ updating: ExampleApp.app.dSYM/Contents/Resources/DWARF/ (stored 0%)
[18:59:41]: ▸ updating: ExampleApp.app.dSYM/Contents/Resources/DWARF/ExampleApp (deflated 90%)
[18:59:41]: ▸ updating: ExampleApp.app.dSYM/Contents/Info.plist (deflated 52%)

[18:59:41]: Successfully exported and compressed dSYM file
[18:59:41]: Successfully exported the .app file:
[18:59:41]: ./fastlane/builds/gamma/ExampleApp
[18:59:41]: Generated plist file with the following values:
[18:59:41]: ▸ -----------------------------------------
[18:59:41]: ▸ {
[18:59:41]: ▸   "method": "development",
[18:59:41]: ▸   "iCloudContainerEnvironment": "Production"
[18:59:41]: ▸ }
[18:59:41]: ▸ -----------------------------------------
[18:59:41]: $ /usr/bin/xcrun /Library/Ruby/Gems/2.6.0/gems/fastlane-2.161.0/gym/lib/assets/wrap_xcodebuild/xcbuild-safe.sh -exportArchive -exportOptionsPlist '/var/folders/gh/zjkfmsr50_bbk_3mmcflp_bm0000gn/T/gym_config20200930-33227-9c8ntw.plist' -archivePath /Users/.../Library/Developer/Xcode/Archives/2020-09-30/ExampleApp\ 2020-09-30\ 18.56.39.xcarchive -exportPath '/var/folders/gh/zjkfmsr50_bbk_3mmcflp_bm0000gn/T/gym_output20200930-33227-16vmy8t' -allowProvisioningUpdates 
[18:59:47]: The generated archive is invalid, this can have various reasons:
[18:59:47]: Usually it's caused by the `Skip Install` option in Xcode, set it to `NO`
[18:59:47]: For more information visit https://developer.apple.com/library/ios/technotes/tn2215/_index.html
[18:59:47]: Also, make sure to have a valid code signing identity and provisioning profile installed
[18:59:47]: Follow this guide to setup code signing https://docs.fastlane.tools/codesigning/GettingStarted/
[18:59:47]: If your intention was only to export an ipa be sure to provide a valid archive at the archive path.
[18:59:47]: This error might also happen if your workspace/project file is not in the root directory of your project.
[18:59:47]: To workaround that issue, you can wrap your calls to gym with
[18:59:47]: `Dir.chdir('../path/to/dir/containing/proj') do`
[18:59:47]: For an example you can check out
[18:59:47]: https://github.com/artsy/emission-nebula/commit/44fe51a7fea8f7d52f0f77d6c3084827fe5dd59e
```

If I add:

```ruby
        skip_package_pkg: true
```

then it succeeds, but the resulting app is using the `Development` iCloud environment.

I'm still poking around, but the problem seems to be in `xcbuild-safe.sh`. If I pass the *exact* same command-line arguments directly to `xcbuild` rather than to `xcbuild-safe.sh` then a) the export succeeds and b) the exported app uses the `Production` iCloud environment.

**Comment by michaellenaghan** [comment]
> I'm still poking around, but the problem seems to be in xcbuild-safe.sh. If I pass the exact same command-line arguments directly to xcbuild rather than to xcbuild-safe.sh then a) the export succeeds and b) the exported app uses the Production iCloud environment.

No, that's not right. In fact, if I use the original command-line, including `xcbuild-safe.sh`, it succeeds:

```
/usr/bin/xcrun /Library/Ruby/Gems/2.6.0/gems/fastlane-2.161.0/gym/lib/assets/wrap_xcodebuild/xcbuild-safe.sh -exportArchive -exportOptionsPlist '/var/folders/gh/zjkfmsr50_bbk_3mmcflp_bm0000gn/T/gym_config20200930-33227-9c8ntw.plist' -archivePath /Users/.../Library/Developer/Xcode/Archives/2020-09-30/ExampleAppApp\ 2020-09-30\ 18.56.39.xcarchive -exportPath '/var/folders/gh/zjkfmsr50_bbk_3mmcflp_bm0000gn/T/gym_output20200930-33227-16vmy8t' -allowProvisioningUpdates 
+ xcodebuild -exportArchive -exportOptionsPlist /var/folders/gh/zjkfmsr50_bbk_3mmcflp_bm0000gn/T/gym_config20200930-33227-9c8ntw.plist -archivePath '/Users/.../Library/Developer/Xcode/Archives/2020-09-30/ExampleApp 2020-09-30 18.56.39.xcarchive' -exportPath /var/folders/gh/zjkfmsr50_bbk_3mmcflp_bm0000gn/T/gym_output20200930-33227-16vmy8t -allowProvisioningUpdates
2020-09-30 19:18:48.674 xcodebuild[34238:6278187] [MT] IDEDistribution: -[IDEDistributionLogging _createLoggingBundleAtPath:]: Created bundle at path '/var/folders/gh/zjkfmsr50_bbk_3mmcflp_bm0000gn/T/Gamma-Release-Scheme_2020-09-30_19-18-48.672.xcdistributionlogs'.
Exported Gamma-Release-Scheme to: /var/folders/gh/zjkfmsr50_bbk_3mmcflp_bm0000gn/T/gym_output20200930-33227-16vmy8t
** EXPORT SUCCEEDED **
```

So the problem seems to be with how whatever follows interprets the result of that command?

**Comment by EpaL** [comment]
This looks like the same issue. When `skip_package_pkg: false`, it will fail because the current Fastlane code only recognize a .pkg output from the build process. If you're building a .app, it will fail because it can't find YourApp.pkg and so the build fails. Adding `skip_package_pkg: true` works around this issue, but means you skip the export process, and your production signing and environment config won't be applied.

**Comment by michaellenaghan** [comment]
I think the error is getting triggered [here](https://github.com/fastlane/fastlane/blob/86f9ea1696fb167b796531f98c409a95bc647450/gym/lib/gym/runner.rb#L144):

```ruby
      if (Dir[BuildCommandGenerator.archive_path + "/*"]).count == 0
        ErrorHandler.handle_empty_archive
      end
```

There's a lot I don't know about here, so I could be off base, but I think the problem is that that isn't where the output goes; it's actually in `-exportPath`, which in the above examples is under `/var/folders/...`.

Still poking...

**Comment by michaellenaghan** [comment]
I don't (yet) know enough about Fastlane internals to fix the problem correctly. I have to move on to other things right now, so here's what I know so far:

 * [this](https://github.com/fastlane/fastlane/blob/86f9ea1696fb167b796531f98c409a95bc647450/gym/lib/gym/runner.rb#L48) line skips what follows if `skip_package_pkg` is `true`. That's why setting it succeeds — and it's also why setting it avoids doing the required work

 * [this](https://github.com/fastlane/fastlane/blob/86f9ea1696fb167b796531f98c409a95bc647450/gym/lib/gym/runner.rb#L50) line "packages" the app — which in this case means producing an `.app` rather than a `.pkg`

 * [this](https://github.com/fastlane/fastlane/blob/86f9ea1696fb167b796531f98c409a95bc647450/gym/lib/gym/runner.rb#L52) line tries to move the `.pkg` and fails

 * the actual failure is at [this](https://github.com/fastlane/fastlane/blob/5e864f8850014f661a557b77b7a199b76cdf4a7d/gym/lib/gym/generators/package_command_generator_xcode7.rb#L102) line; the method that fails is trying to determine the path of the `.pkg`, but of course no `.pkg` exists

It seems like there's a missing `.app` branch in the logic, just as @EpaL was saying. A quick and dirty fix might be to hack `.app` support into [this](https://github.com/fastlane/fastlane/blob/5e864f8850014f661a557b77b7a199b76cdf4a7d/gym/lib/gym/generators/package_command_generator_xcode7.rb#L84) method, but it seems like a proper fix would go in [this](https://github.com/fastlane/fastlane/blob/86f9ea1696fb167b796531f98c409a95bc647450/gym/lib/gym/runner.rb#L16) method.

Well, maybe. :-)

**Comment by joshdholtz** [comment]
@EpaL If you want to a submit a PR for this I’ld be happy to work with you and getting it ready to merge!

**Comment by michaellenaghan** [comment]
For what it's worth...

I took a closer look at this over the weekend and was able to get everything working, at least for my use case. 

My use case is:

 * a Catalyst app 
 * using CloudKit.

The CloudKit part matters because it means that I have to specify an `export_option`, `iCloudContainerEnvironment`, when building. (Well, when "exporting" the build.)

I created these four lanes:

```ruby
# desc 'Build for debug (Mac) (App Store)'
# lane :build_debug_mac_app_store do
#     build_mac_app(
#         export_method: 'app-store',
#         export_options: {
#             iCloudContainerEnvironment: 'Development'  # Not allowed w/ app-store method.
#         },
#         scheme: 'Debug'
#     )
# end

desc 'Build for debug (Mac) (Development)'
lane :build_debug_mac_development do
    build_mac_app(
        export_method: 'development',
        export_options: {
            iCloudContainerEnvironment: 'Development'
        },
        scheme: 'Debug',
        skip_package_pkg: true
    )
end

desc 'Build for release (Mac) (App Store)'
lane :build_release_mac_app_store do
    build_mac_app(
        export_method: 'app-store',
        export_options: {
            iCloudContainerEnvironment: 'Production'
        }
    )
end

desc 'Build for release (Mac) (Development)'
lane :build_release_mac_development do
    build_mac_app(
        export_method: 'development',
        export_options: {
            iCloudContainerEnvironment: 'Production'
        },
        skip_package_pkg: true
    )
end
```

The first two lanes produce `debug` builds, the last two produce `release` builds.

The first and third lanes produce `app-store` builds, the second and fourth produce `development` builds.

The first lane is commented out because `debug` + `app-store` is an invalid combination.

(`app-store` is the default `export_method` and doesn't *really* need to be specified. Also, specifying `export_method` and specifying `method` within `export_options` seem to be equivalent.)

Btw, my `Gymfile` doesn't add much, just a few defaults:

```ruby
clean(true)
scheme('Release')
output_directory('./fastlane/builds')
xcargs('-allowProvisioningUpdates')
```

**Comment by noahsark769** [comment]
@EpaL I'd definitely be interested in seeing your solution. If you don't feel comfortable submitting it for a PR, I'd be happy to do it on your behalf, since it seems like I'm running into the same issue. Thanks!

**Comment by noahsark769** [comment]
I think this is still an issue, correct? Just commenting so the stale not doesn't close it

**Comment by alinradut** [comment]
Yeah, this is still an issue, I tried it yesterday and it still produces an invalid binary if skip_package_pkg is true or build_mac_app crashes with the error above if it's false.

I tried the @EpaL's fork from https://github.com/EpaL/fastlane/commit/6362e6a4c893fe632e4400655851cf9b309c3c2f and indeed it fixes the issue, I think it would be a welcome addition to fastlane if the author agrees.

**Comment by EpaL** [comment]
> I tried the @EpaL's fork from [EpaL@6362e6a](https://github.com/EpaL/fastlane/commit/6362e6a4c893fe632e4400655851cf9b309c3c2f) and indeed it fixes the issue, I think it would be a welcome addition to fastlane if the author agrees.

More than happy for it to be included. I would suggest someone who knows that part of Fastlane give it a once-over first. It works, but I'm sure it could be tidied up a bit.

**Comment by noahsark769** [comment]
@EpaL I think the next step would be to open a PR, since @joshdholtz has already indicated he'd be happy to review it. If it needs to be tidied up, that can probably happen in PR, I think 👍

**Comment by joshdholtz** [comment]
I'd be happy to review a PR 😊

**Comment by EpaL** [comment]
Ok, I'm on it! (Sorry guys, I know I've been a bit slack ;-))

**Comment by appfrilans** [comment]
Following. Any updates on this? Thinking of automating using fastlane, and developer export for macOS (catalyst) is a requirement for me.

**Comment by rogerluan** [comment]
Hey @appfrilans, @EpaL opened a PR with this fix https://github.com/fastlane/fastlane/pull/17885 - I just added testing steps to its description, feel free to test his solution if you can 🤗

**Comment by appfrilans** [comment]
> Hey @appfrilans, @EpaL opened a PR with this fix #17885 - I just added testing steps to its description, feel free to test his solution if you can 🤗

Would love to test. However I'm not sure how to install this PR as a version. Had some major issues just installing my normal fastlane installation.

**Comment by rogerluan** [comment]
@appfrilans you need to use `bundler` (which is generally a good practice, if you're not using bundler yet). Then you can modify your Gemfile as specified in the PR description 😊

**Comment by appfrilans** [comment]
Think my bundler installation is broken. Have also previous spent a lot of time trying to get it to work. Latest error is `ERROR: Failed to build gem native extension.` and similar errors. Sorry but I've given up there.

**Comment by rogerluan** [comment]
Ah, not sure if there're other ways to test it other than using bundler 😬 have you tried completely uninstalling bundler and then installing it again? Maybe it's not a bundler issue after all, it could be some inner dependencies you have in your machine that are conflicting with our gem's dependencies 🤔 not sure though, just wild guessing.

**Comment by appfrilans** [comment]
Have fastlane installed on two computers but both have issues with bundler. Got error like this:

`minitest-5.13.0 requires ruby version ~> 2.2, which is incompatible with the current version, ruby 3.0.0p0`

To bad you have to be a expert in the area not to mess up everything. Cannot test your code, but please let me know when it's part of fastlanes official release!

**Comment by noahsark769** [comment]
Seems like you're using ruby 3.0.0. Using [chruby](https://github.com/postmodern/chruby) to change your ruby to 2.2 might help! Unfortunately ruby versions are notoriously hard to install, change, and manage, so can't promise it will be that easy 😬

**Comment by noahsark769** [comment]
Am I right that this is still an issue? @joshdholtz should we reopen this?

**Comment by rogerluan** [comment]
Sure is @noahsark769 , reopening. Thanks for bringing it up again
